src/base/listener.spec.ts

```diff
@@ -369,4 +388,112 @@ describe('check listening to sockets', function () {
 
     await Promise.all([stopNode(node.listener), stopNode(stunServer)])
   })
+
+  it('add relay node only once', async function () {
+    const stunServer = await startStunServer(undefined, { msgReceived: Defer() })
+
+    const relay = await startNode([new Multiaddr(`/ip4/127.0.0.1/udp/${stunServer.address().port}`)], {
+      msgReceived: Defer()
+    })
+
+    const node = await startNode([new Multiaddr(`/ip4/127.0.0.1/udp/${stunServer.address().port}`)], {
+      msgReceived: Defer()
+    })
+
+    node.publicNodesEmitter.emit(
+      `publicNode`,
+      new Multiaddr(`/ip4/127.0.0.1/tcp/${relay.listener.getPort()}/p2p/${relay.peerId.toB58String()}`)
+    )
+
+    // Let events happen
+    await new Promise((resolve) => setTimeout(resolve))
+    await new Promise((resolve) => setTimeout(resolve))
+    await new Promise((resolve) => setTimeout(resolve))
+    await new Promise((resolve) => setTimeout(resolve))
+
+    let addrs = node.listener.getAddrs().map((ma: Multiaddr) => ma.toString())
+
+    assert(addrs.includes(`/p2p/${relay.peerId.toB58String()}/p2p-circuit/p2p/${node.peerId.toB58String()}`))
+
+    node.publicNodesEmitter.emit(
+      `publicNode`,
+      new Multiaddr(`/ip4/127.0.0.1/tcp/${relay.listener.getPort()}/p2p/${relay.peerId}`)
+    )
+
+    // Let Events happen
+    await new Promise((resolve) => setTimeout(resolve))
+
+    let addrsAfterSecondEvent = node.listener.getAddrs().map((ma: Multiaddr) => ma.toString())
+
+    assert(addrs.length == addrsAfterSecondEvent.length)
+
+    assert(
+      addrsAfterSecondEvent.includes(`/p2p/${relay.peerId.toB58String()}/p2p-circuit/p2p/${node.peerId.toB58String()}`)
+    )
+
+    await Promise.all([stopNode(node.listener), stopNode(relay.listener), stopNode(stunServer)])
+  })
+
+  it('overwrite existing relays', async function () {
+    const stunServer = await startStunServer(undefined, { msgReceived: Defer() })
+
+    const relay = await startNode([new Multiaddr(`/ip4/127.0.0.1/udp/${stunServer.address().port}`)], {
+      msgReceived: Defer()
+    })
+
+    const node = await startNode([new Multiaddr(`/ip4/127.0.0.1/udp/${stunServer.address().port}`)], {
+      msgReceived: Defer()
+    })
+
+    node.publicNodesEmitter.emit(
+      `publicNode`,
+      new Multiaddr(`/ip4/127.0.0.1/tcp/${relay.listener.getPort()}/p2p/${relay.peerId.toB58String()}`)
+    )
+
+    // Let events happen
+    await new Promise((resolve) => setTimeout(resolve))
```

Comment:
somehow I don't like this pattern, can it be done better?

Reply:
fixed, now using proper events